029-across-step/proposal.md

```diff
@@ -0,0 +1,321 @@
+PR: one for Go 1.12, and one for Go 1.13.
+
+### Controlling parallelism with `max_in_flight`
+
+By default, the steps are executed serially to prevent surprising load
+increases from a dynamic var source suddenly returning a ton of values.
+
+To run steps in parallel, a `max_in_flight` must be specified as either `all`
+or a number - its default is `1`. Note: this value is specified on each `var`,
+rather than the entire step.
+
+With `max_in_flight: all`, no limit on parallelism will be enforced. This would
+be typical for when a small, static set of values is specified, and it would be
+annoying to keep the number in sync with the set:
+
+```yaml
+task: unit
+vars: {go-version: ((.:go-version))}
+across:
+- var: go-version
+  values: [1.12, 1.13]
+  max_in_flight: all
+```
+
+With `max_in_flight: 3`, a maximum of 3 var values would be executed in
+parallel. This would be typically set for values coming from a var source,
+which may change at any time, or especially large static values.
+
+```yaml
+set_pipeline: pr
+instance_vars: {pr_number: ((.:pr.number))}
+across:
+- var: pr
+  source: booklit-prs
+  max_in_flight: 3
+```
+
+When multiple `max_in_flight` values are configured, they are multiplicative,
+building on the concurrency of previously listed vars:
+
+```yaml
+set_pipeline: pr
+instance_vars:
+  pr_number: ((.:pr.number))
+  go_version: ((.:go_version))
+across:
+- var: pr
+  source: booklit-prs
+  max_in_flight: 3
+- var: go_version
+  values: [1.12, 1.13]
+  max_in_flight: all
+```
+
+This will run 6 `set_pipeline` steps at a time, focusing on 3 PRs and setting
+Go 1.12 and Go 1.13 pipelines for each in parallel.
+
+Note that setting a `max_in_flight` on a single `var` while leaving the rest as
+their default (`1`) effectively sets an overall max-in-flight.
+
+### Triggering on changes
+
+With `trigger: true` configured on a var, the build plan will run on any change
+to the set of vars - i.e. when a var value is added, removed, or changes.
+
+```yaml
+var_sources:
+- name: booklit-prs
+  type: github-prs
+  config:
+    repository: vito/booklit
+    access_token: # ...
+
+plan:
+- set_pipeline: pr
+  instance_vars: {pr_number: ((.:pr.number))}
+  across:
+  - var: pr
+    source: booklit-prs
+    trigger: true
+```
+
+Note that this can be applied to either static `values:` or dynamic vars from
+`source:` - both cases just boil down to a comparison against the previous
+build's set of values.
+
+### Modifier syntax precedence
+
+The `across` step is a *modifier*, meaning it is attached to another step.
+Other examples of modifiers are `timeout`, `attempts`, `ensure`, and the `on_*`
+family of hooks.
+
+In terms of precedence, `across` would bind more tightly than `ensure` and
+`on_*` hooks, but less tightly than `across` and `timeout`.
+
+`ensure` and `on_*` bind to the `across` step so that they may be run after the
+full matrix completes.
+
+`attempts` binds to the inner step because it doesn't seem to make a whole lot
```

Comment:
@drunkirishcoder going to move this discussion to a review comment thread to avoid cluttering the top-level.

> that can auto retry x attempts. but would there be a manual way? like in the concourse UI rerun the task with the same settings but only retry the failed combo?
>
> the use case I'm trying to design for is I want to use `across` to deploy to multiple clusters/environments. and if one of the env failed, I wouldn't want to rerun the deployment to the successful ones again. so ideally if I can rerun the same matrix, but only for the failed ones again.


Ah, so that's what you meant. There are no plans for this currently, and it would be a bit challenging to get right - there's no precedence for partial reruns of a build. In general, we'd still need to run everything before the `across` step again, since those steps may produce artifacts that the `across` step relies on (e.g. `get`s or `task` outputs).

I think the ideal solution would for the deploy steps to be idempotent - i.e. given the same inputs, if the deployment already succeeded the step would no-op, so re-running the whole matrix wouldn't be an issue. That's perhaps easier said than done, though.

Is there a reason why manual retries is preferable to using `attempts` for automated retries?

Reply:
@aoldershaw ok, thank you for the detailed response. that makes a lot of sense. as to why prefer manual over automated retries. I'm just thinking sometimes a deployment would fail due to circumstances that may be longer to recover, like someone forgot to apply a firewall rule. but yeah I understand why it would be difficult to do in concourse. thanks.

Reply:
One thing you could do is configure a pipeline for each cell in the matrix using the `set_pipeline` step, rather than setting up all the environments within the single build. That way you could retrigger the failed job in each pipeline.